### PR TITLE
Azure Cloud Services Wrapper

### PR DESCRIPTION
**Pull request by ynigoreyes**
**Purpose:** I want to be able to set up the cluster to where when it is running, on whatever dev space, that they do not have to supply a connection string of any sort to any of the services. Having to manually input connection strings to put them in user-secrets sucks and is not production-friendly.

**Solution:** Key Vault will be our best friend along with Azure Flex Volume. I will be using a Service Principal to authenticate the cluster with Azure AD so that the cluster itself can read the secrets available in Key Vault.

**Problems:**

So Secrets are only available to the pods running in the same namespace. Since our development environment could consist of `n` number of namespaces, we need a way to handle the provisioning of secrets to all of those namespaces.

**Option 1:**
We have scripts to basically copy and deploy the secrets already provisioned in the default namespace. The only thing is that we would need a script available to both Windows and Unix systems. This would take place when they run `devsetup.sh`.

**Option 2: (OPTION OF CHOICE)**
We would have a central `SecretsService` deployed in the default namespace so that we are able to have a single source of truth for connection strings and other secrets that only the cluster is able to access through HTTP(S) calls. This reduces the need to copy and deploy the default secrets to all the other namespaces, however... I just don't know how complex the networking is to communicate across namespaces.

Screen Shot 2019-09-21 at 2 56 58 PM
Screen Shot 2019-09-21 at 2 57 13 PM

### CODE REVIEW

**Comment by swoldemi** [comment]
Option 2. I think it's much better to have a single service that interrogates Azure Key Vault for secrets. Also less work for onboarding. Copying a Secret resource across namespaces doesn't seem as clean, I guess. (I think this is what you meant by Option1?)

Pods not in the namespace that the SecretsService pod is running can still make requests to it, not complex: https://stackoverflow.com/questions/37221483/service-located-in-another-namespace

I can do a small test to verify this if you want later.

**Comment by ynigoreyes** [comment]
> Option 2. I think it's much better to have a single service that interrogates Azure Key Vault for secrets. Also less work for onboarding. Copying a Secret resource across namespaces doesn't seem as clean, I guess. (I think this is what you meant by Option1?)
> 
> Pods not in the namespace that the SecretsService pod is running can still make requests to it, not complex: https://stackoverflow.com/questions/37221483/service-located-in-another-namespace
> 
> I can do a small test to verify this if you want later.

Yeah the copying and deploying was a hack that someone came up with. A small test would actually be extremely helpful. I'm learning A LOT about not only kubernetes, but literally the entire Azure Stack so all guidance is welcome :) Thank you

@swoldemi

File: src/SecretsService/Providers/EnvironmentReader.cs [hunk 1]
```diff
@@ -1,29 +1,26 @@
...
 namespace ACMTTU.NoteSharing.SecretsService.Providers
 {
     public static class EnvironmentReader
     {
-        public static EnvironmentVariablePayload GetEnvironmentVariableFromKeyVault(EnvironmentVariableOptions option)
+        public static EnvironmentVariablePayload GetEnvironmentVariableFromKeyVault(ClientOptions option)
         {
             EnvironmentVariablePayload payload = new EnvironmentVariablePayload();
 
             switch (option)
```
Comment (swoldemi) [inline src/SecretsService/Providers/EnvironmentReader.cs:13]: Does the service return the credentials for ALL stages to the client?

Also the exception that's thrown when the `default:` case is hit should be more like $"Invalid resource requested: {option}"

Edit: Looks like it's filtered out in https://github.com/ACMTTU/acm-note-sharing-platform/pull/14/commits/900ac8eacdf7d13866b46abdef6bca7e626272ac#diff-ead3c0ace27db71ecda8b29deae742c4R19 👌

**Review by swoldemi — COMMENTED** [review]
Just 1 thing for you to confirm 👇

File: src/SecretsService/Providers/EnvironmentReader.cs (hunk 1 above)
Comment (ynigoreyes) [reply src/SecretsService/Providers/EnvironmentReader.cs:13]: yes, I thought it would be easier to return all 3 environment variables and have the SDK decide

Comment (ynigoreyes) [reply src/SecretsService/Providers/EnvironmentReader.cs:13]: And yeah that default case should never get hit. That's why I say something really weird happend

Comment (swoldemi) [reply src/SecretsService/Providers/EnvironmentReader.cs:13]: Yeah, that sounds good to me then. Do clients need to decode the base64 secret when using the SDK then? Just saw your Postman screenshots.

Comment (ynigoreyes) [reply src/SecretsService/Providers/EnvironmentReader.cs:13]: All the services that want to use a database just need to do this...

```csharp
   DatabaseClientFactory dbClientFactory = new DatabaseClientFactory()
   CosmosClient client = await dbClientFactory.GetClientAsync()
```

But the more i think about it... I should really be using dependency injection

```csharp
class Application {
  private ClientFactory clientFactory;

  public Application(IClientFactory clientFactory) {
    this.clientFactory = clientFactory;
  }

  public async dynamic FindAllTheDataInDatabase() {
    using (CosmosClient databaseClient = await this.clientFactory.GetClientAsync<ClientOptions.Database>()) {
      // Do whatever needs to get done with the database
    }
    
  }
}
```

**Comment by ynigoreyes** [comment]
@zachlefevre I'm not really sure how to automate tests for this... I feel like there is no point in try to test this section on it's own because it does not provide any confidence that it actually works. The whole point of this library is so that we can grab clients really easily and have the API call abstracted completely

File: src/SecretsService/Providers/EnvironmentReader.cs (hunk 1 above)
Comment (swoldemi) [reply src/SecretsService/Providers/EnvironmentReader.cs:13]: Both work so I'll let you decide on what you'd like everyone else to do. Other than that, LGTM 👍

**Comment by ynigoreyes** [comment]
@zachlefevre I added that one FakeHttpFactory thingy to this PR. I needed to use it too so i put it in Shared/TestingUtils

File: src/SecretsService/Providers/EnvironmentReader.cs (hunk 1 above)
Comment (zachlefevre) [reply src/SecretsService/Providers/EnvironmentReader.cs:13]: DI is the way to go if you want to mock your DB to test your DB logic.

File: src/SecretsService/Startup.cs [hunk 2]
```diff
@@ -0,0 +1,40 @@
+﻿using Microsoft.AspNetCore.Builder;
+using Microsoft.AspNetCore.Hosting;
+using Microsoft.AspNetCore.Mvc;
+using Microsoft.Extensions.Configuration;
+using Microsoft.Extensions.DependencyInjection;
+
+namespace ACMTTU.NoteSharing.SecretsService
+{
+    public class Startup
```
Comment (zachlefevre) [inline src/SecretsService/Startup.cs:10]: I think we should be injecting the DB, HTTP Web Client, web client validation middleware and Neptune services at startup.

File: src/Shared/Sdk/Providers/ServiceClients/ClientFactory.cs [hunk 3]
```diff
@@ -0,0 +1,119 @@
...
+        /// <summary>
+        /// Calls the Secrets Service to grab the specified connection strings
+        /// </summary>
+        /// <param name="option">Your choice of whether you want the Database or the Storage Client</param>
+        /// <returns></returns>
+        public async Task<string> GetConnectionStringForClient(ClientOptions option)
+        {
+            string connectionString;
+
+            HttpClient client = this.factory.CreateClient();
```
Comment (zachlefevre) [inline src/Shared/Sdk/Providers/ServiceClients/ClientFactory.cs:39]: Without a using statement the HttpClient returned by this factory will never close unless explicitly closed. The factory will return a new instance each time creating a memory leak.

File: src/Shared/Sdk/Providers/ServiceClients/ClientOptions.cs [hunk 4]
```diff
@@ -0,0 +1,12 @@
+namespace ACMTTU.NoteSharing.Shared.SDK.Clients
+{
+    /// EnvironmentVariableOptions
+    /// 
+    /// Determines what environment variables are available
+    /// for services in Platform to consume
+    public enum ClientOptions
```
Comment (zachlefevre) [inline src/Shared/Sdk/Providers/ServiceClients/ClientOptions.cs]: I'm generally not a fan of having an enum of a set of options like this, and would instead like to see BlobStorage and Database be classes polymorphic to one another, and both inherit from the same base. This would clean up the case statements above, as well as allow easy mocking for testing purposes. It also means new ClientOptions can be created without having to keep a running list of all options up to date.

File: src/Shared/Sdk/Providers/ServiceClients/IClientFactory.cs [hunk 5]
```diff
@@ -0,0 +1,14 @@
...
+using System.Threading.Tasks;
+using ACMTTU.NoteSharing.Shared.DataContracts;
+using Microsoft.Azure.Cosmos;
+using Microsoft.Azure.Storage.Blob;
+
+namespace ACMTTU.NoteSharing.Shared.SDK.Clients
+{
+    public interface IClientFactory
+    {
+        CloudBlobClient GetStorageClient(string connectionString);
```
Comment (zachlefevre) [inline src/Shared/Sdk/Providers/ServiceClients/IClientFactory.cs]: Small suggestion: Do not use string for the connectionString, instead use the ConnectionString class in the stdlib. This makes parsing of the string unnecessary, it allows you to select individual fields easily, and it avoids having a "string-ly typed" typeclass.

**Review by zachlefevre — CHANGES_REQUESTED** [review]
I would suggest breaking up a PR this large into a set of smaller commits:
https://graysonkoonce.com/stacked-pull-requests-keeping-github-diffs-small/

Overall this looks really good! You obviously put a lot of work into ensuring the people on the project have a smooth and realistic project experience. There are only a few things I noticed that could be changed. I am unfamiliar enough with the helm work that I suggest somebody else review it.

File: src/SecretsService/Providers/EnvironmentReader.cs (hunk 1 above)
Comment (ynigoreyes) [reply src/SecretsService/Providers/EnvironmentReader.cs:13]: Yes I did that :)

File: src/SecretsService/Startup.cs (hunk 2 above)
Comment (ynigoreyes) [reply src/SecretsService/Startup.cs:10]: Well... we have the chicken and the egg problem. Neither of our PRs are in so we I can't do that just yet

File: src/Shared/Sdk/Providers/ServiceClients/IClientFactory.cs (hunk 5 above)
Comment (ynigoreyes) [reply src/Shared/Sdk/Providers/ServiceClients/IClientFactory.cs]: I never have to parse it. 

I got this code directly from the [Quickstart](https://docs.microsoft.com/en-us/azure/storage/blobs/storage-quickstart-blobs-dotnet)

It looks like this is the recommended way for connecting to the database/storage accounts

File: src/Shared/Sdk/Providers/ServiceClients/ClientOptions.cs (hunk 4 above)
Comment (ynigoreyes) [reply src/Shared/Sdk/Providers/ServiceClients/ClientOptions.cs]: That was my first implementation actually so I 100% agree with you. I would rather have two different types of client factories. One for the CosmosDB account and one for the Storage account.

The reason why you don't see it on here, is because I would then have one method...

```csharp
public ClientType GetClient<ClientType>(string connectionString) {
  return Client
}
```

I don't know how to return either a CosmosDB or a Storage client from a method, much less create an interface for it.

I'm starting to think...

Do you think I'm trying to do too much with this class lib?

Right now, I am getting the connection string, as well as creating a Client. Do you think I should leave it at grabbing the connection string and leave the creation of the Client itself to the users?

Comment (ynigoreyes) [reply src/Shared/Sdk/Providers/ServiceClients/ClientOptions.cs]: well actually... hold on lemme work on this real quick. But I do agree with you.

File: src/Shared/Sdk/Providers/ServiceClients/ClientFactory.cs (hunk 3 above)
Comment (ynigoreyes) [reply src/Shared/Sdk/Providers/ServiceClients/ClientFactory.cs:39]: I am away of the IDispoable interface and that we do use using...

I actually did some research on this and there are a few things I found

https://aspnetmonsters.com/2016/08/2016-08-27-httpclientwrong/

TL;DR look at the bottom of the article

Comment (ynigoreyes) [reply src/Shared/Sdk/Providers/ServiceClients/ClientFactory.cs:39]: https://stackoverflow.com/questions/15705092/do-httpclient-and-httpclienthandler-have-to-be-disposed

Another one.

File: src/SecretsService/Startup.cs (hunk 2 above)
Comment (ynigoreyes) [reply src/SecretsService/Startup.cs:10]: But yes, when both our PR's are in, we should definitely inject the stuff we just built into all the services and applications

File: src/Shared/Sdk/Providers/ServiceClients/ClientOptions.cs (hunk 4 above)
Comment (zachlefevre) [reply src/Shared/Sdk/Providers/ServiceClients/ClientOptions.cs]: Look at the IHttpClientFactory: The method CreateClient() is the method required to return the constructed client. Then you would have

```chsarp
public static DBClient CreateClient(ClientType t, ConnectionString cs){
...
}
```
If you use the factory pattern you can leave fetching to the user and allow construction to the DI IoC container.

File: src/Shared/Sdk/Providers/ServiceClients/ClientFactory.cs (hunk 3 above)
Comment (zachlefevre) [reply src/Shared/Sdk/Providers/ServiceClients/ClientFactory.cs:39]: I see... How strange. In that case we need to ensure we use a singleton service, not a transient one, when using DI.

File: src/Shared/Sdk/Providers/ServiceClients/IClientFactory.cs (hunk 5 above)
Comment (zachlefevre) [reply src/Shared/Sdk/Providers/ServiceClients/IClientFactory.cs]: I was wrong on what the class was. It is a connectionstring builder: https://docs.microsoft.com/en-us/dotnet/framework/data/adonet/connection-string-builders.

Very strange that there is an EventHubConnectionString, but not one to connect to a CosmosDB: https://docs.microsoft.com/en-us/azure/event-hubs/event-hubs-dotnet-standard-getstarted-send

I wonder if the SqlCSB works. Unsure. Either way it's not a big deal.

**Review by zachlefevre — DISMISSED** [review]

File: src/Shared/Sdk/Providers/ServiceClients/StorageClientFactory.cs [hunk 6]
```diff
@@ -0,0 +1,32 @@
...
+namespace ACMTTU.NoteSharing.Shared.SDK.Clients
+{
+    public class StorageClientFactory : ClientFactory
+    {
+        public StorageClientFactory(IHttpClientFactory factory) : base(factory)
+        {
+            this.secretServiceUrl = new Uri(this.secretServiceUrl.AbsoluteUri + "Database");
+        }
+
+        public async Task<CloudBlobClient> GetClient(string connectionString)
```
Comment (zachlefevre) [inline src/Shared/Sdk/Providers/ServiceClients/StorageClientFactory.cs]: GetClient does not override anything on the base class and therefore on any location where a ClientFactory is required GetClient cannot be called.

The fundamental problem is that the two db types are not polymorphic to one another, and therefore expressing in the type system that a method should return one or the other, when they do not share a base methods is impossible.

What we need instead is some kind of ad-hoc polymorphism OR structural polymorphism preferably in the form of type classes.

Comment (zachlefevre) [reply src/Shared/Sdk/Providers/ServiceClients/StorageClientFactory.cs]: Actually it's late and I'm overcomplicating this. The problem is still the same, but the solution is to implement a simple generic factory. Changes pushed.

**Review by zachlefevre — DISMISSED** [review]

File: src/SecretsService/Startup.cs (hunk 2 above)
Comment (zachlefevre) [reply src/SecretsService/Startup.cs:10]: Sry, getting ahead of myself. We can't the validation middleware, but we can the IHttpClientFactory and the database.

File: src/Shared/Sdk/Providers/ServiceClients/StorageClientFactory.cs (hunk 6 above)
Comment (zachlefevre) [reply src/Shared/Sdk/Providers/ServiceClients/StorageClientFactory.cs]: Note: This does not damage type safety in any way. Any service which needs a blob or relational service will accept in it's constructor parameter the ClientFactory<whatever> where whatever is the type of storage it needs.

I think we can make this cleaner if the ClientFactory<T>.CreateClient<T> => DBClient<T>. DBClient<T> would then contain all of the logic pertaining to using that storage mechanism. This class would service as an anti corruption layer.
This is a bad idea if we intend to use all of the mechanisms of each DB technology whenever the DbClient<T> is used.

What Would Dijkstra Do?

Comment (ynigoreyes) [reply src/Shared/Sdk/Providers/ServiceClients/StorageClientFactory.cs]: OoooooOoooOoo I like the changes. But how does that I create our type safety?

Comment (zachlefevre) [reply src/Shared/Sdk/Providers/ServiceClients/StorageClientFactory.cs]: ?

**Comment by zachlefevre** [comment]
Note for anybody who is new to the (beautiful) factory pattern: This factory is useful for constructing ANY type whose construction involves the (secretsServiceName OR the secretsServiceName) AND an IHttpClientFactory.
Because the two types, the blob storage and relational storage types are not in the same type hierarchy (and therefore not polymorphic to one another) the generic must be T with no type constraints.

File: src/Shared/Sdk/Providers/ServiceClients/StorageClientFactory.cs (hunk 6 above)
Comment (ynigoreyes) [reply src/Shared/Sdk/Providers/ServiceClients/StorageClientFactory.cs]: Nvm i read it and figured it out

**Comment by ynigoreyes** [comment]
> I would suggest breaking up a PR this large into a set of smaller commits:
> https://graysonkoonce.com/stacked-pull-requests-keeping-github-diffs-small/
> 
> Overall this looks really good! You obviously put a lot of work into ensuring the people on the project have a smooth and realistic project experience. There are only a few things I noticed that could be changed. I am unfamiliar enough with the helm work that I suggest somebody else review it.

I don't really know how to break it up since it has already been commited :/

**Comment by ynigoreyes** [comment]
Is it good to go in yet or is there is there still things?

@zachlefevre

**Review by zachlefevre — APPROVED** [review]
